contributor-docs/adrs/adr-003-prop-norms.md

```diff
@@ -0,0 +1,53 @@
+# ADR 003: Prop norms in Primer React Components
+
+## Status
+
+In progress / draft / pending?
+
+## Context
+
+- PRC APIs matter
+- Consistency is important
+- There’s a gap between where we want to be and where we
+- This document aims to codify our best practices for component APIs and unify disparate conversations
+  - In the course of some refactorings, we may have intermediate states where there will added be value in a clear documentation of which parts of our API reflect the best practices.
+
+## Decision
+
+### 🟢 `sx`
+
+All components that ultimately render to the DOM should accept an `sx` prop. The `sx` prop (of type `SystemStyleObject`) should set styles for the root HTML element rendered by the component. It may make sense to have an additional prop to style a descendant element (e.g., `innerSx`).
+
+### 🟢 `ref`
+
+All components that ultimately render to the DOM should accept a `ref` prop. That `ref` prop should be passed to the root HTMLElement rendered by the component.
+
+### 🟡 `as`: Limited
+
+Only components with a clear need for polymorphism should accept an `as` prop. Reasonable cases include:
+
+- Components that need functionality from the `as` prop, like a button
+- Components whose accessibility are improved by using semantically appropriate HTML elements
```

Comment:
The reasoning behind when a component should accept the `as` prop is a little unclear here.

I think we should also mention restricting the allowed values for the `as` prop. Some questions to answer:
* When do should we allow anything to be passed to the `as` prop?
* When should the possible values be limited?
* How do we decide on a limited set of values?

Reply:
Oh interesting. I had assumed that _if_ we ever have an `as` prop, we would allow any `ElementType`.

Reply:
Primer ViewComponents restricts it for some components: https://github.com/github/primer/discussions/130#discussioncomment-676056